File: 백준/Gold/7576. 토마토/토마토.js
const fs = require('fs')
const input = fs.readFileSync(0, 'utf8').trim().split('\n');
const [M, N] = input[0].split(" ").map(Number)
const visited = Array.from({ length: N }, () => Array(M).fill(false));
const board = input.slice(1).map(item => item.split(" ").map(Number))

const dx = [0, -1, 0, 1]
const dy = [-1, 0, 1, 0]

const queue = []
let result = 1;
let head = 0

for (let j = 0; j < N; j++) {
    for (let i = 0; i < M; i++) {
        if (board[j][i] === 1) {
            queue.push([j, i, 1])
            visited[j][i] = true
        }
    }
}
const bfs = () => {
    while (head < queue.length) {
        const [y, x, day] = queue[head++]
        for (let i = 0; i < 4; i++) {
            const next_x = x + dx[i]
            const next_y = y + dy[i]
            if (
                next_x < 0 || next_x >= M ||
                next_y < 0 || next_y >= N
            ) continue
            if (board[next_y][next_x] === -1) continue
            if (visited[next_y][next_x]) continue
            queue.push([next_y, next_x, day + 1])
            visited[next_y][next_x] = true
            board[next_y][next_x] = day + 1 
            result = Math.max(result,day + 1)
        }
    }
}

bfs()


for (let j = 0; j < N; j++) {
    for (let i = 0; i < M; i++) {
        if (board[j][i] === 0 ) {
            console.log(-1)
            return
        }
    }
}

console.log(result - 1)

File: 백준/Gold/7576. 토마토/토마토.py
from collections import deque
M, N = map(int, input().split())

board = []
for i in range(N):
    board.append(list(map(int, input().split())))

directions = [(0, 1), (-1, 0), (0, -1), (1, 0)]


def bfs():
    queue = deque()
    tomato_arr = []
    for i in range(N):
        for j in range(M):
            if board[i][j] == 1:
                queue.append((i, j))
    while queue:
        currY, currX = queue.popleft()
        for dy, dx in directions:
            ny = currY + dy
            nx = currX + dx
            if ny < 0 or ny >= N or nx < 0 or nx >= M:
                continue
            if board[ny][nx] == 0:
                board[ny][nx] = board[currY][currX] + 1
                queue.append((ny, nx))


bfs()

for i in range(N):
    for j in range(M):
        if board[i][j] == 0:
            print(-1)
            exit()

result = 0
for i in range(N):
    for j in range(M):
        result = max(result, board[i][j])
print(result - 1)
